webapi/setaltsig.go

```diff
@@ -0,0 +1,137 @@
+
+	// Get values which have been added to context by middleware.
+	sigAddress := c.MustGet("SigningAddress").(string)
+	dcrdClient := c.MustGet("DcrdClient").(Node)
+	reqBytes := c.MustGet("RequestBytes").([]byte)
+
+	if cfg.VspClosed {
+		sendError(errVspClosed, c)
+		return
+	}
+
+	var request SetAltSigRequest
+	if err := binding.JSON.BindBody(reqBytes, &request); err != nil {
+		log.Tracef("%s: Bad request (clientIP=%s): %v", funcName, c.ClientIP(), err)
+		sendErrorWithMsg(err.Error(), errBadRequest, c)
+		return
+	}
+
+	altPubAddr, ticketHash := request.AltPubAddress, request.TicketHash
+
+	currentAltSigAddr, err := db.AltSigAddr(ticketHash)
+	if err != nil {
+		log.Errorf("%s: db.AltSigExists (ticketHash=%s): %v", funcName, ticketHash, err)
+		sendError(errInternalError, c)
+		return
+	}
+	if currentAltSigAddr != "" {
+		msg := "alternate signature address already exists"
+		log.Errorf("%s: %s (ticketHash=%s)", funcName, msg, ticketHash)
+		sendErrorWithMsg(msg, errBadRequest, c)
+		return
+
+	}
+
+	// Disallow empty addresses.
+	if altPubAddr == "" {
+		const details = "Cannot add new blank alt sig address"
+		log.Tracef("%s: %s (clientIP=%s): %v", funcName, details, c.ClientIP(), err)
+		sendErrorWithMsg(details, errBadRequest, c)
+		return
+	}
+
+	// Fail fast if the pubkey doesn't decode properly.
+	addr, err := stdaddr.DecodeAddressV0(altPubAddr, cfg.NetParams)
+	if err != nil {
+		log.Tracef("%s: Alt pub address cannot be decoded (clientIP=%s): %v", funcName, c.ClientIP(), err)
+		sendErrorWithMsg(err.Error(), errBadRequest, c)
+		return
+	}
+	if _, ok := addr.(*stdaddr.AddressPubKeyHashEcdsaSecp256k1V0); !ok {
+		log.Tracef("%s: Alt pub address is unexpected type (clientIP=%s type=%T): %v", funcName, c.ClientIP(), addr, err)
+		sendErrorWithMsg("wrong type for alternate signing address", errBadRequest, c)
+		return
+	}
+
+	// Get ticket details.
+	rawTicket, err := dcrdClient.GetRawTransaction(ticketHash)
+	if err != nil {
+		log.Errorf("%s: dcrd.GetRawTransaction for ticket failed (ticketHash=%s): %v", funcName, ticketHash, err)
+		sendError(errInternalError, c)
+		return
+	}
+
+	// Ensure this ticket is eligible to vote at some point in the future.
+	canVote, err := dcrdClient.CanTicketVote(rawTicket, ticketHash, cfg.NetParams)
+	if err != nil {
+		log.Errorf("%s: dcrd.CanTicketVote error (ticketHash=%s): %v", funcName, ticketHash, err)
+		sendError(errInternalError, c)
+		return
+	}
+	if !canVote {
+		log.Tracef("%s: unvotable ticket (clientIP=%s, ticketHash=%s)",
+			funcName, c.ClientIP(), ticketHash)
+		sendError(errTicketCannotVote, c)
+		return
+	}
+
+	sigStr := c.GetHeader("VSP-Client-Signature")
+	sig, err := base64.StdEncoding.DecodeString(sigStr)
+	if err != nil {
+		log.Tracef("%s: error decoding signature string (ticketHash=%s, sig=%s): %v", funcName, ticketHash, sigStr, err)
+		sendError(errInternalError, c)
+		return
+	}
+	history := &database.AltSigHistory{
+		Addr: sigAddress,
+		Req:  reqBytes,
```

Comment:
Thoughts on recording response and response signature here, as we do for vote change records?

Reply:
That hadn't occurred to me as something necessary. We could always spoof that later, right? We have the keys. Does it make something more secure here? Will add though.

Reply:
9aebd65 and 0175871

Reply:
You're correct to say that a server can forge a response at will, at any time.

Hmm. Trying to imagine a scenario where it would make sense.

1. Client sends a request to set voting preferences (X). Server encounters an error, does not update database, provides an error response to client.
1. As a result, server votes with wrong preferences (Y).
1. Client makes a complaint.
1. Admin investigates. In their DB they see the request which set the X preferences, and they know that their server voted Y, but they have no way of knowing that the response was an error. Confusion and/or argument ensue.

Maybe this is a bit contrived? I think we originally justified storing the response by thinking that a complete audit trail would be more valuable than just half the trail, and the cost of storing the responses is relatively cheap.

Reply:
Yeah, it's probably good accounting. No problem saving it. Just wondering if there was something important I was missing.